<!DOCTYPE html>
<html>
<head>
<title>
<!--Title-->
</title>
<style>
TABLE {width: 80%; border-collapse: collapse; border: 2px solid black;}
TD, TH {border: 1px solid grey; padding: 4px;}
CAPTION {caption-side: bottom; font-size: small; padding-top: 10px; padding-bottom: 30px;}
</style>
<body>
<center><h1>
<!--Title-->
</h1>
<table>
<caption><b>Table 1.</b> Number of aligned reads and open bases in each sample. <i>Factor</i> is a scaling factor
computed based on the number of aligned reads and on the total open bases. <i>Peaks</i> is the number of peaks
detected in each sample. <i>FRIP</i> is the fraction of reads that fall in peaks. According to ENCODE guidelines,
this should be over 30%, and definitely not less than 20%.</caption>
<!--Table1-->
</table>
<br>
<table>
<caption><b>Table 2.</b> WashU genome browser tracks for all samples, and for all conditions.</caption>
<!--Table2-->
</table>
<br><br>
<table>
<caption><b>Table 3.</b> Number of comparable peaks and average peak sizes in each contrast.</caption>
<!--Table3-->
</table>
<br>
<table>
<caption><b>Table 4.</b> Differential peaks in each contrast. To determine differential peaks, the peaks in both conditions (test and control)
are merged, to form a set of comparable peaks. The number of aligned reads in each alignable peak is then computed for all replicates of the
two conditions, to produce a peak size matrix. Finally, this matrix is analyzed with DESeq2 to identify peaks that are significantly higher
in the test or in the control condition. The <i>Significant</i> column reports the number of comparable peaks showing a significant difference
between test and control; the <i>Up in Test</i> and <i>Up in Ctrl</i> columns show the number of peaks that are higher in test and in the
control condition respectively. The Excel files in the <i>Peaks</i> column contain the position, fold change, and P-value for all
significantly different peaks (Sig:) and for all peaks (All:).</caption>
<!--Table4-->
</table>
<br>
<table>
<caption><b>Table 5.</b> Genes showing increased or decreased accessibility at transcription start sites.</caption>
<!--Table5-->
</table>
<br>
<table>
<caption><b>Table 6.</b> Genes showing increased or decreased accessibility measured across the whole gene body.</caption>
<!--Table6-->
</table>
<br>
<table>
<caption><b>Table 7.</b> Enhancers showing increased or decreased accessibility.</caption>
<!--Table7-->
</table>
<br>
<table>
<caption><b>Table 8.</b> Plots based on contrast data. The plots in the <i>Peak sizes</i> column show a scatterplot of the size of common peaks
in the Test and Control conditions.  Dots above the diagonal indicate that the peak is larger in the Test sample than in the Ctrl sample. <i>TSS</i>
plots show the ATAC profile around transcription start sites in the Test and Control conditions. <i>Test Peaks</i> heatmaps show the profile of
peaks that are significantly higher in Test (from Table 3) in both conditions. <i>Ctrl Peaks</i> shows the same information for peaks that are 
higher in the Control condition.</caption>
<!--Table8-->
</table>
<br>

</center>
</body>
</html>
